File: editors/code/src/lsp_ext.ts
/**
 * This file mirrors `src/lsp_ext.rs` declarations.
 */

import * as lc from "vscode-languageclient";

export const preprocessedDocument = new lc.RequestType<
  PreprocessedDocumentParams,
  string,
  void
>("sourcepawn-studio/preprocessedDocument");

export type PreprocessedDocumentParams = {
  textDocument?: lc.TextDocumentIdentifier;
};

export const syntaxTree = new lc.RequestType<SyntaxTreeParams, string, void>(
  "sourcepawn-studio/syntaxTree"
);

export type SyntaxTreeParams = {
  textDocument?: lc.TextDocumentIdentifier;
};

export const analyzerStatus = new lc.RequestType<
  AnalyzerStatusParams,
  string,
  void
>("sourcepawn-studio/analyzerStatus");

export type AnalyzerStatusParams = {
  textDocument?: lc.TextDocumentIdentifier;
};

export const itemTree = new lc.RequestType<ItemTreeParams, string, void>(
  "sourcepawn-studio/itemTree"
);

export type ItemTreeParams = {
  textDocument?: lc.TextDocumentIdentifier;
};

export const projectMainPath = new lc.RequestType<
  ProjectMainPathParams,
  lc.URI,
  void
>("sourcepawn-studio/projectMainPath");

export type ProjectMainPathParams = {
  uri?: lc.URI;
};

export const projectsGraphviz = new lc.RequestType<
  ProjectsGraphvizParams,
  string,
  void
>("sourcepawn-studio/projectsGraphviz");

export type ProjectsGraphvizParams = {
  textDocument?: lc.TextDocumentIdentifier;
};

export const serverStatus = new lc.NotificationType<ServerStatusParams>(
  "sourcepawn-studio/serverStatus"
);
export type ServerStatusParams = {
  health: "ok" | "warning" | "error";
  quiescent: boolean;
  message?: string;
};

export const spcompStatus = new lc.NotificationType<ServerStatusParams>(
  "sourcepawn-studio/spcompStatus"
);
export type SpcompStatusParams = {
  quiescent: boolean;
};

export const hover = new lc.RequestType<
  lc.HoverParams,
  (lc.Hover & { actions: CommandLinkGroup[] }) | null,
  void
>(lc.HoverRequest.method);

export type CommandLink = {
  /**
   * A tooltip for the command, when represented in the UI.
   */
  tooltip?: string;
} & lc.Command;
export type CommandLinkGroup = {
  title?: string;
  commands: CommandLink[];
};

File: crates/sourcepawn_lsp/src/lsp_ext.rs
use lsp_types::{notification::Notification, request::Request, TextDocumentIdentifier, Url};
use serde::{Deserialize, Serialize};

pub enum PreprocessedDocument {}

impl Request for PreprocessedDocument {
    type Params = PreprocessedDocumentParams;
    type Result = String;
    const METHOD: &'static str = "sourcepawn-lsp/preprocessedDocument";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PreprocessedDocumentParams {
    pub text_document: Option<TextDocumentIdentifier>,
}

pub enum SyntaxTree {}

impl Request for SyntaxTree {
    type Params = SyntaxTreeParams;
    type Result = String;
    const METHOD: &'static str = "sourcepawn-lsp/syntaxTree";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct SyntaxTreeParams {
    pub text_document: Option<TextDocumentIdentifier>,
}

pub enum ItemTree {}

impl Request for ItemTree {
    type Params = ItemTreeParams;
    type Result = String;
    const METHOD: &'static str = "sourcepawn-lsp/itemTree";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ItemTreeParams {
    pub text_document: Option<TextDocumentIdentifier>,
}

pub enum ProjectMainPath {}

impl Request for ProjectMainPath {
    type Params = ProjectMainPathParams;
    type Result = Url;
    const METHOD: &'static str = "sourcepawn-lsp/projectMainPath";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ProjectMainPathParams {
    pub uri: Option<Url>,
}

pub enum ProjectsGraphviz {}

impl Request for ProjectsGraphviz {
    type Params = ProjectsGraphvizParams;
    type Result = String;
    const METHOD: &'static str = "sourcepawn-lsp/projectsGraphviz";
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct ProjectsGraphvizParams {
    pub text_document: Option<TextDocumentIdentifier>,
}

pub enum ServerStatusNotification {}

impl Notification for ServerStatusNotification {
    type Params = ServerStatusParams;
    const METHOD: &'static str = "sourcepawn-lsp/serverStatus";
}

#[derive(Deserialize, Serialize, PartialEq, Eq, Clone, Debug)]
pub struct ServerStatusParams {
    pub health: Health,
    pub quiescent: bool,
    pub message: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Copy, PartialEq, Eq, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Health {
    Ok,
    Warning,
    Error,
}

pub enum SpcompStatusNotification {}

impl Notification for SpcompStatusNotification {
    type Params = SpcompStatusParams;
    const METHOD: &'static str = "sourcepawn-lsp/spcompStatus";
}

#[derive(Deserialize, Serialize, PartialEq, Eq, Clone, Debug)]
pub struct SpcompStatusParams {
    pub quiescent: bool,
}
